### PR TITLE
Modernize & secure demo version installation scripts

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
Let's do some practice in scripting:

* Migrate demo installation script for Windows entirely to PowerShell (since we are using PowerShell commands in BAT file anyway);
* Get rid of 7zip requirement on Windows, use built-in shell support for ZIP archives instead;
* Make just one `demo_unix.sh` script instead of `demo_linux.sh` and `demo_macos.sh` with autodetection for wget/curl (macOS doesn't have wget by default and AFAIK Ubuntu doesn't have curl by default);
* Last, but not least: verify SHA256 checksums for downloaded files and abort installation on mismatch.

P.S.

* (Optional) Migrate windows/install_packages.bat from 7zip to PowerShell's Expand-Archive (available since PowerShell 5.0, Windows 8+)

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
`Get-FileHash` is available since PowerShell 4.0, so it should be available on Windows 7 SP1 and later.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , Windows script can't be run on Windows 7:
```
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

d:\fheroes2\fheroes2\script\demo>demo_windows.bat
[1/4] downloading demo version
FATAL ERROR: The term 'Get-FileHash' is not recognized as the name of a cmdlet,
function, script file, or operable program. Check the spelling of the name, or i
f a path was included, verify that the path is correct and try again.
Press any key to exit...
```

My powershell version is **2.0**.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub Well, it's sad :( I'll make an exception handler for the case if Get-FileHash isn't supported.

**Comment by oleg-derevenetz** [comment]
@ihhub BTW should this script ever run in unattended mode via AppVeyor? BAT file had corresponding check at the end, but I don't see why it needed.

File: script/windows/install_packages.bat [hunk 1]
```diff
@@ -24,49 +24,36 @@ echo [7/7] Downloading packages
...
-where 7z.exe >nul 2>nul
-if %errorlevel% == 0 (
-    set sevenZipPath=7z.exe
-) else (
-    if exist "%ProgramFiles%\7-Zip\7z.exe" (
-        set sevenZipPath=%ProgramFiles%\7-Zip\7z.exe
-    )
-)
+echo [1/7] Unpacking packages
+powershell -Command "Expand-Archive -LiteralPath 'zlib\zlib1.2.11.zip' -DestinationPath 'zlib' -Force"
```
Comment (ihhub) [inline script/windows/install_packages.bat:31]: `Expand-Archive` is not supported on PowerShell 2.0. Could we please use the same trick as for demo downloading?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , nice work! I left few more comments here. Would you mind to take a look?

File: script/windows/install_packages.bat (hunk 1 above)
Comment (oleg-derevenetz) [reply script/windows/install_packages.bat:31]: Hi @ihhub

> Could we please use the same trick as for demo downloading?

The problem is that this script is intended to run in AppVeyor as well, and that trick uses the shell (explorer.exe), but I highly doubt that Windows VM does have it running. That's why I asked if demo_windows.bat should ever be used with AppVeyor or not. From the other side, this script is for developers, not end users, may be we can sacrifice a PowerShell 2.0 compatibility in this case?

Comment (ihhub) [reply script/windows/install_packages.bat:31]: demo_windows.bat could be used in the future for unit testing but for now it's not used anywhere. Regarding install_packages.bat script: at least I have to update my powershell😅

Comment (oleg-derevenetz) [reply script/windows/install_packages.bat:31]: Hi @ihhub We can make a separate versions then: for users/developers (with shell) and for AppVeyor (with Expand-Archive). What do you think?

Comment (ihhub) [reply script/windows/install_packages.bat:31]: Hi @oleg-derevenetz , looks like a good idea! Let's do this.

File: script/demo/demo_unix.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+set -e
+
+H2DEMO_URL="https://archive.org/download/HeroesofMightandMagicIITheSuccessionWars_1020/h2demo.zip"
+H2DEMO_SHA256="12048c8b03875c81e69534a3813aaf6340975e77b762dc1b79a4ff5514240e3c"
+
+[[ ! -d demo ]] && mkdir demo
+cd demo
+
+if [[ "$(which wget 2> /dev/null)" != "" ]]; then
```
Comment (qarkai) [inline script/demo/demo_unix.sh:11]: Please use bash builtin `command -v` instead of `which`.

Comment (oleg-derevenetz) [reply script/demo/demo_unix.sh:11]: Hi @qarkai What's the benefit (apart from using built-in command itself)?

Comment (qarkai) [reply script/demo/demo_unix.sh:18]: `which` command adds unnecessary dependency on package `which`.

**Comment by qarkai** [comment]
Does this PR supersedes #3455 and #3456?

File: script/demo/demo_unix.sh (hunk 2 above)
Comment (oleg-derevenetz) [reply script/demo/demo_unix.sh:18]: @qarkai  That's a good reason. I'll replace `which` by `command` then.

Comment (qarkai) [reply script/demo/demo_unix.sh:18]: Also with `command` we can simply check return value, not test output:
`if command -v wget > /dev/null; then echo 'found'; fi`

Comment (oleg-derevenetz) [reply script/demo/demo_unix.sh:18]: > Also with command we can simply check return value, not test output:

Yes, I know. But it doesn't seems much shorter than

`if [[ "$(command -v wget)" != "" ]]; then echo 'found'; fi`

:)

**Comment by oleg-derevenetz** [comment]
@qarkai

> Does this PR supersedes #3455 and #3456?

Yes, it seems so.

File: script/demo/demo_unix.sh (hunk 2 above)
Comment (qarkai) [reply script/demo/demo_unix.sh]: It's not about "shorter". It's about "simpler and faster" since that method doesn't call extra `test` command.

Comment (oleg-derevenetz) [reply script/demo/demo_unix.sh]: `test` as well as `[[` are built-in constructions of bash anyway. "Premature micro optimizations" of such kind are usually useless.

**Comment by ihhub** [comment]
Hi @hgdagon, could you please take a look at this pull request? I think it's the same way as you have in your pull requests. My apologies if we do double job :(

**Comment by hgdagon** [comment]
@ihhub no need to apologize, however it seems like I won't be able to finish my PR. I can't for the life of me figure out how batch scripts work, I had issues with jumps, if statements, comparisons, exiting loops vs exiting the script itself, etc. that no amount of documentation online helped me figure out what's going on. So, if @oleg-derevenetz has a solution, we should definitely go with this one.

That said, from the get-go I was looking into a method to skip the download (since it just takes too long) if the file already exists, so I did actually learn that `Get-Filehash` doesn't work in PS2.0. One solution I found online was `Certutil`, like this

```batchfile
certutil -hashfile demo.zip SHA256 | find /V "hash"
```

I'm also not quite sure about switching to ps scripts entirely, because of the execution policy BS. Here, we're still running from a batch script, which is something I was trying to avoid.

P.S. WOW! 110 lines for downloading a small demo?

**Comment by oleg-derevenetz** [comment]
Hi @hgdagon

> One solution I found online was Certutil, like this
> `certutil -hashfile demo.zip SHA256 | find /V "hash"`

I can try to adopt `certutil` as a fallback variant if `Get-FileHash` is not available, in separate PR. However, due to localized nature of its output, the above command may not work "as is" depending on user language. Need to think about different locale-independent approach.

> I'm also not quite sure about switching to ps scripts entirely, because of the execution policy BS. Here, we're still running from a batch script, which is something I was trying to avoid.

BAT files are utter crap in a modern Windows environment. They have a poor syntax, very weak functionality (without external utilities or inclusions of PowerShell code), insufficient error handling capabilities. PowerShell, on the other hand, lives up to its name in terms of scripting (however, to my taste, it's not so good as the **shell** itself, in comparison with bash or zsh, but that's another story).

**Comment by hgdagon** [comment]
> BAT files are utter crap in a modern Windows environment. They have a poor syntax, very weak functionality 

Yes, but they allows us to run ps commands directly. That's my only concern here. However, I do see the limitations, i.e. if I understand correctly, you can't send variables from one ps command to the other. You would need to take care of variables in batch and then pass them on to the next command.

> due to localized nature of its output, the above command may not work "as is" depending on user language

I see, but if we're sure the hash is always on line 2, maybe we could use this?

```powershell
certutil -hashfile demo.zip SHA256 | select -first 1 -skip 1
```

Also, this reminds me exactly why I wanted to use unzip. It allows us to extract only the 2 directories that we need (unless we're extracting the whole thing for a dev, but that, obviously, is also very easy with unzip) and not having to bother with any copy operations, resulting in a much leaner script.

That said, BSD-tar is also another option on Win10 (don't remember since which build, though).

**Comment by oleg-derevenetz** [comment]
> if I understand correctly, you can't send variables from one ps command to the other. You would need to take care of variables in batch and then pass them on to the next command.

Yes, and there may be complex variables like COM objects and so on. BAT files are completely unaware of such things.

> I see, but if we're sure the hash is always on line 2, maybe we could use this?

I did it in a [slightly different manner](https://github.com/oleg-derevenetz/fheroes2/blob/7d8bb5804d5003e42d651e637fdba417195dd653/script/demo/demo_windows.ps1#L19):

`return (certutil.exe -hashfile $Path sha256 | Select-String -Pattern "^[0-9A-Fa-f]{64}$").Line.ToUpper()`

but I think it's a matter of taste.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , what I received while running `demo_windows.bat`:
```
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

d:\fheroes2\fheroes2\script\demo>demo_windows.bat
[1/4] downloading demo version
FATAL ERROR: You cannot call a method on a null-valued expression.
Press any key to exit...
```

**Comment by oleg-derevenetz** [comment]
Hi @ihhub Do you have certutil installed? What does the `certutil.exe -hashfile h2demo.zip sha256` print?

**Comment by ihhub** [comment]
> Hi @ihhub Do you have certutil installed? What does the `certutil.exe -hashfile h2demo.zip sha256` print?

Here is the output:
```
d:\fheroes2\fheroes2\script\demo\demo>certutil.exe -hashfile h2demo.zip sha256
CertUtil: -hashfile command FAILED: 0xd00000bb (-805306181)
CertUtil: WsResetMetadata
```

**Comment by oleg-derevenetz** [comment]
@ihhub 

> Here is the output:
> 
> ```
> d:\fheroes2\fheroes2\script\demo\demo>certutil.exe -hashfile h2demo.zip sha256
> CertUtil: -hashfile command FAILED: 0xd00000bb (-805306181)
> CertUtil: WsResetMetadata
> ```

OMG. 0xc00000bb is STATUS_NOT_SUPPORTED, interesting, what exactly is "not supported" here. Any suggestions for such cases? Ignore error and print a warning?

**Comment by ihhub** [comment]
> @ihhub
> 
> > Here is the output:
> > ```
> > d:\fheroes2\fheroes2\script\demo\demo>certutil.exe -hashfile h2demo.zip sha256
> > CertUtil: -hashfile command FAILED: 0xd00000bb (-805306181)
> > CertUtil: WsResetMetadata
> > ```
> 
> OMG. 0xc00000bb is STATUS_NOT_SUPPORTED, interesting, what exactly is "not supported" here. Any suggestions for such cases? Ignore error and print a warning?

I think yes, let's just pring the warning. Even despite of fact that this is not recommended. We should have some solution (which is already better than before) right now which we could improve in the future. We shouldn't waste so much time on this.

**Comment by ihhub** [comment]
@oleg-derevenetz , excellent job! Thanks a lot for these changes!

**Pull request merged**